test/test_examples.py

```diff
@@ -1,17 +1,40 @@
 # This should be an absolute import from package base: "gw_spaceheat...."
 # That requires changes *all* imports to use absolute import, so we don't do this in this demo.
 import time
+import typing
+from collections import defaultdict
+
 from data_classes.cacs.temp_sensor_cac import TempSensorCac
 from data_classes.sh_node import ShNode
 import load_house
 from actors.power_meter import PowerMeter
 from actors.primary_scada import PrimaryScada
 from actors.atn import Atn
 from actors.tank_water_temp_sensor import TankWaterTempSensor
+from schema.gs.gs_dispatch import GsDispatch
 from universal_test_ear import UniversalTestEar
 from schema.gt.gt_telemetry.gt_telemetry_maker import GtTelemetry, GtTelemetry_Maker
 from schema.gs.gs_pwr_maker import GsPwr_Maker
-# noinspection PyUnresolvedReferences
+
+
+class ScadaRecorder(PrimaryScada):
+    """Record data about a PrimaryScada execution during test"""
+
+    num_received: int
+    num_received_by_topic: typing.Dict[str, int]
+
+    def __init__(self, node: ShNode):
+        self.num_received = 0
+        self.num_received_by_topic = defaultdict(int)
+        super().__init__(node)
+
+    def on_mqtt_message(self, client, userdata, message):
+        self.num_received += 1
+        self.num_received_by_topic[message.topic] += 1
+        super().on_mqtt_message(client, userdata, message)
+
+    def gs_dispatch_received(self, payload: GsDispatch, from_node: ShNode):
+        raise NotImplementedError
```

Comment:
I've cleaned up the `atn_base.py` and `primary_scada_base.py`, so abstractmethod `gs_dispatch_received` is gone and `on_gw_message` exists. See [https://github.com/thegridelectric/gw-scada-spaceheat-python/blob/jm/async-power-test/gw_spaceheat/actors/primary_scada_base.py#L44] here in the other PR